File: LeetCodePractice/src/DP/LongestPalindromicSubstring.java
package DP;

public class LongestPalindromicSubstring {
	public String longestPalindrome(String s) {
		int length = s.length();
		int begin = 0;
		int maxlen = 1;
		int[][] table = new int[length][length];

		for (int i = 0; i < length; i++) {
			table[i][i] = 1;
		}

		for (int i = 0; i < length - 1; i++) {
			if (s.charAt(i) == s.charAt(i + 1)) {
				table[i][i + 1] = 1;
				maxlen = 2;
				begin = i;
			}
		}

		for (int len = 3; len <= length; len++) {
			for (int i = 0; i + len - 1 < length; i++) {
				int j = i + len - 1;
				if (s.charAt(i) == s.charAt(j) && table[i + 1][j - 1] == 1) {
					table[i][j] = 1;
					if (len > maxlen) {
						maxlen = len;
						begin = i;
					}
				}
			}
		}

		return s.substring(begin, begin + maxlen);
	}
}

File: LeetCodePython/LongestPalindromicSubstring.py
"""
Given a string s, return the longest palindromic substring in s.

Example 1:
Input: s = "babad"
Output: "bab"
Explanation: "aba" is also a valid answer.

Example 2:
Input: s = "cbbd"
Output: "bb"

Constraints:
1 <= s.length <= 1000
s consist of only digits and English letters.

hints:
1 How can we reuse a previously computed palindrome to compute a larger palindrome?
2 If “aba” is a palindrome, is “xabax” a palindrome? Similarly is “xabay” a palindrome?
3 Complexity based hint: If we use brute-force and check whether for every start and end position a substring is a palindrome we have O(n^2) start - end pairs and O(n) palindromic checks.
Can we reduce the time for palindromic checks to O(1) by reusing some previous computation.

analysis:
TC: O(N^2)
"""
class LongestPalindromicSubstring:
    def longestPalindrome(self, s: str) -> str:
        self.max_len, self.start = 0, 0
        if len(s) < 2:
            return s
        for i in range(len(s)):
            self.expand(s, i, i)
            self.expand(s, i, i + 1)
        return s[self.start: self.start + self.max_len]

    def expand(self, s, l, r):
        while l >= 0 and r < len(s) and s[l] == s[r]:
            l -= 1
            r += 1
        if r - l - 1 >= self.max_len:
            self.max_len = r - l - 1
            self.start = l + 1